agent/handlers/types/v2/response.go

```diff
@@ -0,0 +1,152 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package v2
+
+import (
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/containermetadata"
+	"github.com/aws/amazon-ecs-agent/agent/engine/dockerstate"
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+// TaskResponse defines the schema for the task response JSON object
+type TaskResponse struct {
+	Cluster       string
+	TaskARN       string
+	Family        string
+	Version       string
+	DesiredStatus string `json:",omitempty"`
+	KnownStatus   string
+	Containers    []ContainerResponse `json:",omitempty"`
+	Limits        LimitsResponse      `json:",omitempty"`
+}
+
+// ContainerResponse defines the schema for the container response
+// JSON object
+type ContainerResponse struct {
+	ID            string `json:"DockerId"`
+	Name          string
+	DockerName    string
+	Image         string
+	ImageID       string
+	Ports         []PortResponse    `json:",omitempty"`
+	Labels        map[string]string `json:",omitempty"`
+	DesiredStatus string
+	KnownStatus   string
+	ExitCode      *int `json:",omitempty"`
+	Limits        LimitsResponse
+	// TODO StartedAt time.Time
+	Type     string
+	Networks []containermetadata.Network `json:",omitempty"`
+}
+
+// LimitsResponse defines the schema for task/cpu limits response
+// JSON object
+type LimitsResponse struct {
+	CPU    uint
+	Memory uint
+}
+
+// PortResponse defines the schema for portmapping response JSON
+// object
+type PortResponse struct {
+	ContainerPort uint16
+	Protocol      string
+	HostPort      uint16 `json:",omitempty"`
+}
+
+// NewTaskResponse creates a new response object for the task
+func NewTaskResponse(taskARN string,
+	state dockerstate.TaskEngineState,
+	cluster string) (*TaskResponse, error) {
+	task, ok := state.TaskByArn(taskARN)
+	if !ok {
+		return nil, errors.Errorf("v2 task response: unable to find task '%s'", taskARN)
+	}
+
+	resp := &TaskResponse{
+		Cluster:       cluster,
+		TaskARN:       task.Arn,
+		Family:        task.Family,
+		Version:       task.Version,
+		DesiredStatus: task.GetDesiredStatus().String(),
+		KnownStatus:   task.GetKnownStatus().String(),
+	}
+
+	containerNameToDockerContainer, ok := state.ContainerMapByArn(task.Arn)
+	if !ok {
+		seelog.Warnf("V2 task response: unable to get container name mapping for task '%s'",
+			task.Arn)
+		return resp, nil
+	}
+
+	eni := task.GetTaskENI()
+	for _, dockerContainer := range containerNameToDockerContainer {
```

Comment:
So I understand, this looks like we're exposing the container metadata that the agent is already holding in engine. Is it possible for it to be be stale? Specifically `KnownStatus`, `DesiredStatus`.

Reply:
It is possible for these to be stale. But, we're exposing the state of things as ECS knows them. These should eventually get updated in a normal scenario.